File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def searchArr(self,nums: List[int], target: int,s: int,e: int) -> int:
        if(s>e):
            return -1
        mid = s+(e-s)//2
        if(target==nums[mid]):
            return mid
        if(nums[s] <= nums[mid]):
            if(nums[s] <= target and target <= nums[mid]) :
                return self.searchArr(nums,target,s,mid-1)
            else:
                return self.searchArr(nums,target,mid+1,e)
        
        else:
            if (nums[mid] <= target and target<=nums[e]):
                return self.searchArr(nums,target,mid+1,e)
            else:
                return self.searchArr(nums,target,s,mid-1)
    def search(self, nums: List[int], target: int) -> int:
        s=0
        e=len(nums)-1
        return self.searchArr(nums, target,s,e)
    
    

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int s=0;
        int e=nums.length-1;
        return searching(nums,target,s,e);
    }
    public int searching(int[] nums,int target,int s,int e){
        if(s>e) return -1;
        int mid=(s+e)/2;
        if(target==nums[mid]) return mid;
        if(nums[s]<=nums[mid]){
            if(nums[s]<=target && target<=nums[mid]){
                return searching(nums,target,s,mid-1);
            } else{
                return searching(nums,target,mid+1,e);
            }
        }
        else{
            if(nums[mid]<=target && target<=nums[e]){
                return searching(nums,target,mid+1,e);
            } else{
                return searching(nums,target,s,mid-1);
            }
        }
        
    }
    
    
}